src/core/assistant-message/parseAssistantMessageV3.ts

```diff
@@ -0,0 +1,296 @@
+import { PartialXMLStreamParser, xmlObjectToString } from "../parser"
+import { ToolParamName } from "../../shared/tools"
+import { AssistantMessageContent } from "./parseAssistantMessage"
+import { ToolName, toolNames } from "@roo-code/types"
+
+export function parseAssistantMessageV3(assistantMessage: string): AssistantMessageContent[] {
+	const textNodeNameKey = "#text"
+	const attributeNamePrefixKey = "@"
+	const parser = new PartialXMLStreamParser({
+		textNodeName: textNodeNameKey,
+		attributeNamePrefix: attributeNamePrefixKey,
+		alwaysCreateTextNode: true,
+		parsePrimitives: false,
+		// Removing maxDepth to allow proper parsing of nested structures
+		// maxDepth: 1, // this for backward compatible with old version (old version only support 2 levels (1 level depth))
+		stopNodes: [
```

Comment:
I am not sure if it should be part of this pull request or part of a tool class refactor, but having tool-specific language in a generic parser implementation somehow seems inappropriate.

It would be nice if each tool can register what it expects to be a stopNode

Reply:
also if you need to iterate `stopNodes` regularly, then it should probably be a set or a dictionary so that you do not have to do a linear scan. 

for example, if you have to check `stopNodes` for every new token response from the model while it streams, then a linear scan (`O(n)`) will be slower compared to a dictionary or set lookup ( `O(log(n))` ).